connection.go

```diff
@@ -0,0 +1,596 @@
+	ProducerRequiredAcks                      string        `json:"producer_required_acks"`
+	ProducerTimeout                           time.Duration `json:"producer_timeout"`
+	ProducerRetryMax                          int           `json:"producer_retry_max"`
+	ProducerRetryBackoff                      time.Duration `json:"producer_retry_backoff"`
+	ProducerReturnErrors                      bool          `json:"producer_return_errors"`
+	ProducerReturnSuccesses                   bool          `json:"producer_return_successes"`
+	ProducerFlushMaxMessages                  int           `json:"producer_flush_max_messages"`
+	ProducerCompressionLevel                  string        `json:"producer_compression_level"`
+	ConsumerFetchMin                          int32         `json:"consumer_fetch_min"`
+	ConsumerFetchDefault                      int32         `json:"consumer_fetch_default"`
+	ConsumerRetryBackOff                      time.Duration `json:"consumer_retry_backoff"`
+	ConsumerMaxWaitTime                       time.Duration `json:"consumer_max_wait_time"`
+	ConsumerMaxProcessingTime                 time.Duration `json:"consumer_mex_processing_time"`
+	ConsumerReturnErrors                      bool          `json:"consumer_return_errors"`
+	ConsumerOffsetAutoCommitEnable            bool          `json:"consumer_offset_auto_commit_enable"`
+	ConsumerOffsetAutoCommintInterval         time.Duration `json:"consumer_offset_auto_commit_interval"`
+	ConsumerOffsetsInitial                    int           `json:"consumer_offsets_initial"`
+	ConsumerOffsetsRetryMax                   int           `json:"consumer_offsets_retry_max"`
+	ConsumerGroupSessionTimeout               time.Duration `json:"consumer_group_session_timeout"`
+	ConsumerGroupHeartBeatInterval            time.Duration `json:"consumer_group_heart_beat_interval"`
+	ConsumerGroupRebalanceTimeout             time.Duration `json:"consumer_group_rebalance_timeout"`
+	ConsumerGroupRebalanceRetryMax            int           `json:"consumer_group_rebalance_retry_max"`
+	ConsumerGroupRebalanceRetryBackOff        time.Duration `json:"consumer_group_rebalance_retry_back_off"`
+	ConsumerGroupRebalanceResetInvalidOffsets bool          `json:"consumer_group_rebalance_reset_invalid_offsets"`
+}
+
+type Client struct {
+	ClientID              int
+	AccountName           string
+	NatsConnectionID      string
+	IsConsumer            bool
+	IsProducer            bool
+	LearningFactor        int
+	LearningFactorCounter int
+	LearningRequestSent   bool
+	GetSchemaRequestSent  bool
+	BrokerConnection      *nats.Conn
+	JSContext             nats.JetStreamContext
+	ProducerProtoDesc     protoreflect.MessageDescriptor
+	ProducerSchemaID      string
+	ConsumerProtoDescMap  map[string]protoreflect.MessageDescriptor
+	Counters              ClientCounters
+	Config                Config
+}
+
+type ClientCounters struct {
+	TotalBytesBeforeProduce           int64
```

Comment:
let's make it single counter for before_reduction (before produce/after consume) and after_reduction(after produce/before consume)